### ISSUE TITLE
Docker daemon crash after using the command docker logs and docker service logs

### ISSUE THREAD

**Issue by yuval-p** [moby/moby#35050]
We have docker swarm with dameon version 17.06.
We configured some services with log-opt to use several labels

 for example:

docker service create --name log-test --log-driver json-file --log-opt labels=bla.name.com,capitalString,bla2.name.com,aaa,bbb --container-label bla.name.com=label1 --container-label capitalString=True --container-label bla2.name.com=label2 --container-label aaa=aaa --container-label bbb=bbb --log-opt env=ENVE -e ENVE=12 grafana/grafana

after configuring the service, and using the command:

docker service logs -f log-test --details

the damon crashes..

**Steps to reproduce the issue:**
1. install 17.06
2. docker service create --name log-test --log-driver json-file --log-opt labels=bla.name.com,capitalString,bla2.name.com,aaa,bbb --container-label bla.name.com=label1 --container-label capitalString=True --container-label bla2.name.com=label2 --container-label aaa=aaa --container-label bbb=bbb --log-opt env=ENVE -e ENVE=12 grafana/grafana 
3. docker service logs -f log-test --details

**Describe the results you received:**

sometimes I received the desired logs and sometimes "unexpected EOF"

**Describe the results you expected:**
The service logs

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

Client:
 Version:      17.06.1-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   874a737
 Built:        Thu Aug 17 22:53:49 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.1-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   874a737
 Built:        Thu Aug 17 23:01:50 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

```

**Output of `docker info`:**

```
Containers: 22
 Running: 6
 Paused: 0
 Stopped: 16
Images: 12
Server Version: 17.06.1-ce
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: 5kvv1ow27racgwtw4nj49dufo
 Is Manager: true
 ClusterID: u48fva9t3af9eeuu9w6hg91w8
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: 192.168.0.153
 Manager Addresses:
  192.168.0.153:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.3 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.339GiB
Name: localhost.localdomain
ID: GRWT:JVBR:YJVC:TPLR:ONSE:RVE7:ZPGI:KB4L:UPRY:LPIJ:EQOR:NQBY
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

It happens on several environments, bare metal and virtual machine 

The daemon logs:

Oct  1 10:19:30 localhost dockerd: goroutine 12580 [running]:
Oct  1 10:19:30 localhost dockerd: runtime.throw(0x1a4d584, 0x26)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/runtime/panic.go:596 +0x95 fp=0xc4211ebbc8 sp=0xc4211ebba8
Oct  1 10:19:30 localhost dockerd: runtime.mapiternext(0xc4211ebdd8)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/runtime/hashmap.go:737 +0x7ee fp=0xc4211ebc78 sp=0xc4211ebbc8
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/daemon/cluster/executor/container.(*controller).Logs(0xc4200ffb80, 0x7f5afc623128, 0xc421214540, 0x262a520, 0xc4213ae480, 0xc420978ae0, 0x2, 0x2, 0x1, 0x0, ...)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/cluster/executor/container/controller.go:530 +0x4fe fp=0xc4211ebeb8 sp=0xc4211ebc78
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/agent.(*taskManager).Logs(0xc4203e2a20, 0x7f5afc623128, 0xc421214540, 0xc420978ae0, 0x2, 0x2, 0x1, 0x0, 0x0, 0x262a520, ...)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/agent/task.go:72 +0x110 fp=0xc4211ebf40 sp=0xc4211ebeb8
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/github.com/docker/swarmkit/agent.(*worker).Subscribe.func2(0xc420978b90, 0x7f5afc624478, 0xc420c50e40, 0xc4215badb0, 0x262a520, 0xc4213ae480, 0xc4203e2a20)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/agent/worker.go:544 +0xac fp=0xc4211ebfa8 sp=0xc4211ebf40
Oct  1 10:19:30 localhost dockerd: runtime.goexit()
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc4211ebfb0 sp=0xc4211ebfa8
Oct  1 10:19:30 localhost dockerd: created by github.com/docker/docker/vendor/github.com/docker/swarmkit/agent.(*worker).Subscribe
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/agent/worker.go:545 +0x874
Oct  1 10:19:30 localhost dockerd: goroutine 1 [chan receive, 490 minutes]:
Oct  1 10:19:30 localhost dockerd: main.(*DaemonCli).start(0xc4204395f0, 0x0, 0x1a2db34, 0x17, 0xc420092000, 0xc42043e7c0, 0xc42043a3f0, 0x0, 0x0)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/cmd/dockerd/daemon.go:294 +0x1d25
Oct  1 10:19:30 localhost dockerd: main.runDaemon(0x0, 0x1a2db34, 0x17, 0xc420092000, 0xc42043e7c0, 0xc42043a3f0, 0x0, 0x0)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/cmd/dockerd/docker.go:91 +0x8c
Oct  1 10:19:30 localhost dockerd: main.newDaemonCommand.func1(0xc42008c480, 0x26ae7f0, 0x0, 0x0, 0x0, 0x0)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/cmd/dockerd/docker.go:42 +0x6f
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).execute(0xc42008c480, 0xc4200101c0, 0x0, 0x0, 0xc42008c480, 0xc4200101c0)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:646 +0x44e
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc42008c480, 0x17d4ee0, 0x1, 0xc4204299b0)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:742 +0x349
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).Execute(0xc42008c480, 0xc4204299b0, 0x0)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:695 +0x2b
Oct  1 10:19:30 localhost dockerd: main.main()
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/cmd/dockerd/docker.go:118 +0xe6
Oct  1 10:19:30 localhost dockerd: goroutine 17 [syscall, 490 minutes, locked to thread]:
Oct  1 10:19:30 localhost dockerd: runtime.goexit()
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/runtime/asm_amd64.s:2197 +0x1
Oct  1 10:19:30 localhost dockerd: goroutine 31 [IO wait, 490 minutes]:
Oct  1 10:19:30 localhost dockerd: net.runtime_pollWait(0x7f5afc593cc0, 0x72, 0x0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/runtime/netpoll.go:164 +0x59
Oct  1 10:19:30 localhost dockerd: net.(*pollDesc).wait(0xc420127e98, 0x72, 0x0, 0xc4201b84a0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/fd_poll_runtime.go:75 +0x38
Oct  1 10:19:30 localhost dockerd: net.(*pollDesc).waitRead(0xc420127e98, 0xffffffffffffffff, 0x0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/fd_poll_runtime.go:80 +0x34
Oct  1 10:19:30 localhost dockerd: net.(*netFD).accept(0xc420127e30, 0x0, 0x2628020, 0xc4201b84a0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/fd_unix.go:430 +0x1e5
Oct  1 10:19:30 localhost dockerd: net.(*UnixListener).accept(0xc420260ab0, 0xc4201bac30, 0x177cac0, 0x26010f0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/unixsock_posix.go:162 +0x32
Oct  1 10:19:30 localhost dockerd: net.(*UnixListener).Accept(0xc420260ab0, 0xc4201bac00, 0x177cac0, 0x26010f0, 0x18bb620)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/unixsock.go:237 +0x49
Oct  1 10:19:30 localhost dockerd: net/http.(*Server).Serve(0xc42009e370, 0x263dc60, 0xc420260ab0, 0x0, 0x0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/http/server.go:2643 +0x228
Oct  1 10:19:30 localhost dockerd: net/http.Serve(0x263dc60, 0xc420260ab0, 0x26278a0, 0xc420260ae0, 0x437978, 0x1a9a1a0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/http/server.go:2307 +0x78
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/daemon.(*Daemon).listenMetricsSock.func1(0x263dc60, 0xc420260ab0, 0xc420260ae0)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/metrics_unix.go:31 +0x4b
Oct  1 10:19:30 localhost dockerd: created by github.com/docker/docker/daemon.(*Daemon).listenMetricsSock
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/metrics_unix.go:32 +0x196
Oct  1 10:19:30 localhost dockerd: goroutine 6 [syscall]:
Oct  1 10:19:30 localhost dockerd: os/signal.signal_recv(0x2633f20)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/runtime/sigqueue.go:116 +0x104
Oct  1 10:19:30 localhost dockerd: os/signal.loop()
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/os/signal/signal_unix.go:22 +0x22
Oct  1 10:19:30 localhost dockerd: created by os/signal.init.1
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/os/signal/signal_unix.go:28 +0x41
Oct  1 10:19:30 localhost dockerd: goroutine 11 [syscall, 490 minutes]:
Oct  1 10:19:30 localhost dockerd: syscall.Syscall6(0xf7, 0x1, 0x71cb, 0xc42013fdb0, 0x1000004, 0x0, 0x0, 0x0, 0x0, 0x0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/syscall/asm_linux_amd64.s:44 +0x5
Oct  1 10:19:30 localhost dockerd: os.(*Process).blockUntilWaitable(0xc420260450, 0x0, 0x0, 0x0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/os/wait_waitid.go:28 +0xa5
Oct  1 10:19:30 localhost dockerd: os.(*Process).wait(0xc420260450, 0x0, 0x0, 0x0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/os/exec_unix.go:22 +0x4d
Oct  1 10:19:30 localhost dockerd: os.(*Process).Wait(0xc420260450, 0x0, 0xc420416b60, 0x1a9a468)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/os/exec.go:115 +0x2b
Oct  1 10:19:30 localhost dockerd: os/exec.(*Cmd).Wait(0xc4204d14a0, 0x1a9a1a0, 0x0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/os/exec/exec.go:435 +0x62
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/libcontainerd.(*remote).runContainerdDaemon.func1(0xc4204d14a0, 0xc420079380)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/libcontainerd/remote_unix.go:479 +0x2b
Oct  1 10:19:30 localhost dockerd: created by github.com/docker/docker/libcontainerd.(*remote).runContainerdDaemon
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/libcontainerd/remote_unix.go:481 +0xb7c
Oct  1 10:19:30 localhost dockerd: goroutine 14 [chan receive]:
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/libcontainerd.(*remote).handleConnectionChange(0xc420079380)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/libcontainerd/remote_unix.go:187 +0xe2
Oct  1 10:19:30 localhost dockerd: created by github.com/docker/docker/libcontainerd.New
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/libcontainerd/remote_unix.go:130 +0x651
Oct  1 10:19:30 localhost dockerd: goroutine 13 [select, 490 minutes]:
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/google.golang.org/grpc.(*addrConn).transportMonitor(0xc4204d1600)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/google.golang.org/grpc/clientconn.go:751 +0x666
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/google.golang.org/grpc.(*ClientConn).resetAddrConn.func1(0xc4204d1600)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/google.golang.org/grpc/clientconn.go:506 +0x1d9
Oct  1 10:19:30 localhost dockerd: created by github.com/docker/docker/vendor/google.golang.org/grpc.(*ClientConn).resetAddrConn
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/google.golang.org/grpc/clientconn.go:507 +0x372
Oct  1 10:19:30 localhost dockerd: goroutine 16 [IO wait]:
Oct  1 10:19:30 localhost dockerd: net.runtime_pollWait(0x7f5afc593d80, 0x72, 0x7)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/runtime/netpoll.go:164 +0x59
Oct  1 10:19:30 localhost dockerd: net.(*pollDesc).wait(0xc420127958, 0x72, 0x262da20, 0x2621460)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/fd_poll_runtime.go:75 +0x38
Oct  1 10:19:30 localhost dockerd: net.(*pollDesc).waitRead(0xc420127958, 0xc420208000, 0x8000)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/fd_poll_runtime.go:80 +0x34
Oct  1 10:19:30 localhost dockerd: net.(*netFD).Read(0xc4201278f0, 0xc420208000, 0x8000, 0x8000, 0x0, 0x262da20, 0x2621460)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/fd_unix.go:250 +0x1b7
Oct  1 10:19:30 localhost dockerd: net.(*conn).Read(0xc42042c308, 0xc420208000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/net/net.go:181 +0x70
Oct  1 10:19:30 localhost dockerd: bufio.(*Reader).Read(0xc4204f72c0, 0xc4203de858, 0x9, 0x9, 0x0, 0x0, 0x1)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/bufio/bufio.go:213 +0x312
Oct  1 10:19:30 localhost dockerd: io.ReadAtLeast(0x2623220, 0xc4204f72c0, 0xc4203de858, 0x9, 0x9, 0x9, 0x26270e0, 0xc421a6fbc0, 0xc421a6fbc0)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/io/io.go:307 +0xa9
Oct  1 10:19:30 localhost dockerd: io.ReadFull(0x2623220, 0xc4204f72c0, 0xc4203de858, 0x9, 0x9, 0xc420010070, 0x105000000000002, 0x2)
Oct  1 10:19:30 localhost dockerd: /usr/local/go/src/io/io.go:325 +0x58
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/golang.org/x/net/http2.readFrameHeader(0xc4203de858, 0x9, 0x9, 0x2623220, 0xc4204f72c0, 0x0, 0x0, 0x0, 0x26239e0)
Oct  1 10:19:30 localhost systemd: docker.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/golang.org/x/net/http2/frame.go:237 +0x7b
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/golang.org/x/net/http2.(*Framer).ReadFrame(0xc4203de820, 0x0, 0x0, 0x0, 0x0)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/golang.org/x/net/http2/frame.go:469 +0xa4
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/google.golang.org/grpc/transport.(*framer).readFrame(0xc4202606c0, 0xc421a6fb60, 0xc421a6fb60, 0x0, 0x0)
Oct  1 10:19:30 localhost dockerd: /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/google.golang.org/grpc/transport/http_util.go:508 +0x2f
Oct  1 10:19:30 localhost dockerd: github.com/docker/docker/vendor/google.golang.org/grpc/transport.(*http2Client).reader(0xc4204f9c20)
....

**Comment by thaJeztah** [comment]
@cpuguy83 @aaronlehmann is this the same issue as https://github.com/moby/moby/issues/32418 ?

**Comment by aaronlehmann** [comment]
No, #32418 was a cleanup.

ping @dperny

**Comment by thaJeztah** [comment]
oh, meant to link https://github.com/docker/for-linux/issues/92, which was fixed through https://github.com/moby/moby/pull/34174

**Comment by aaronlehmann** [comment]
Looks to me like it matches that one.

**Comment by thaJeztah** [comment]
That issue should be fixed in 17.09 @yuval-p are you still able to reproduce this on the current 17.09 release?

**Comment by yuval-p** [comment]
@thaJeztah Looks like the issue is fixed in 17.09. 
thanks!
